### PR TITLE
#217: use @props + t-comb for prop evaluation (closes #217)

### LINKED ISSUE #217: use @props + t-comb for prop evaluation
We're using `tcomb` + `tcomb-react` as the standard validators in our projects. We should do the same here :)

### PR DESCRIPTION
**Pull request by FrancescoCioria**
Issue #217

### CODE REVIEW

File: src/dropdown/Dropdown.js [hunk 1]
```diff
@@ -9,22 +10,16 @@ const themes = {
...
-    React.PropTypes.node,
-    React.PropTypes.element,
-    React.PropTypes.array
-  ]),
-  theme: React.PropTypes.oneOf(['semantic']),
-  valueLink: React.PropTypes.shape({
-    value: React.PropTypes.string,
-    requestChange: React.PropTypes.func
-  })
+  children: t.ReactNode,
```
Comment (giogonzo) [inline src/dropdown/Dropdown.js]: just noting that: `children: oneOfType()` was not `.isRequired` before, now it is (probably better anyway)

Comment (FrancescoCioria) [inline src/dropdown/Dropdown.js]: Mmm no it's a mistake thanks (you can use `options` as this is a wrapper of `react-select`)

File: src/dropdown/Dropdown.js [hunk 2]
```diff
@@ -9,22 +10,16 @@ const themes = {
...
+  theme: t.maybe(t.enums.of(['semantic'])),
+  valueLink: t.maybe(t.struct({
+    value: t.maybe(t.String),
+    requestChange: t.Function
+  }))
 };
-
+@props(PropTypes, { strict: false })
```
Comment (giogonzo) [inline src/dropdown/Dropdown.js]: we should add something like "you can pass any other prop and react-select it will be passed to react-select as-is"?

**Comment by giogonzo** [comment]
looks great except for comment :+1: 

one more general comment/question is: how many different situations/strategies we have when dealing with "non strict" prop typing? I mean:
- "general props" case (style, className, id, ?): they are listed in `@props` with no docs, as it is added automatically (no `strict: false` needed)
- "wrapped component api" case: we wrap a first or third party component in another one, but we want to keep the underlying component api accessible in a flat fashion (`strict: false` is needed). this is the case of e.g. `DropDown` and also `TransitionWrapper` I guess (it can wrap any component)
- something else I'm missing?

File: src/dropdown/Dropdown.js (hunk 2 above)
Comment (FrancescoCioria) [inline src/dropdown/Dropdown.js]: yes, definitely. I was talking with @gabro about this issue two days ago: are `docgen` should identify `{ strict: false}` and also parse eventual comment describing **why** props are not strict.
PS: I'm going to add this comment when working on the *"add inline docs"* part of #204

**Comment by FrancescoCioria** [comment]
> "general props" case

I think we should find a better way to handle these three standard props (`style`, `className`, `id`)
First of all, currently, not every component accepts all of them and I think they should, do you agree?

also:
> they are listed in @props with no docs, as it is added automatically (no strict: false needed)

this is not true for the new `revenge-docgen` made by @gabro (but I'll soon open an issue when I'll be working on docs here ;) )

> something else I'm missing?

there may be some cases (`FlexView`) where you want to let user set event listeners on your component (`onClick`, `onMouseEnter`....)

**Comment by giogonzo** [comment]
> First of all, currently, not every component accepts all of them and I think they should, do you agree?

Yes

> this is not true for the new revenge-docgen made by @gabro (but I'll soon open an issue when I'll be working on docs here ;) )

got it, makes sense

> there may be some cases (FlexView) where you want to let users set event listeners on your component (onClick, onMouseEnter....)
PS: this might be considered a "wrapped component api" for div

Yup, although I'd also value keeping apis small and the "do one single thing right" approach: event handlers can probably always be placed in a wrapper or content of a `FlexView`. This same reasoning actually also applies to other general props (`id`, `className`, `style`), it's a just a trade-off between simplicity of implementation and simplicity of usage
> I think we should find a better way to handle these three standard props (style, className, id)

If and when we have a standard way of dealing with "other props" we can apply it to `FlexView` and other components

**Pull request merged**